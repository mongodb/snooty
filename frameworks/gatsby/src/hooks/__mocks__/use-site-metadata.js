module.exports = {
  useSiteMetadata: jest.fn().mockReturnValue({
    branch: 'master',
    project: 'datalake',
    title: 'MongoDB Datalake',
    user: 'andrew',
  }),
};
