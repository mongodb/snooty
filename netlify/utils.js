import fs from 'fs';
import path from 'path';

/**
 * Parses build hook data from the expected temporary location
 * @returns {object | undefined}
 */
const parseBuildHookData = () => {
  // This file does not currently exist, but should be created on Netlify builds as part of `npm run build:netlify`.
  // The INCOMING_HOOK_BODY env var is not automatically passed along to functions, so we use a txt file to save it
  const relativeFilePath = '../build-hook.txt';
  const buildHookDataString = fs.readFileSync(path.resolve(__dirname, relativeFilePath), 'utf-8');
  if (!buildHookDataString) {
    console.log('No build hook data found.');
    return;
  }
  return JSON.parse(buildHookDataString);
};

export const constructResPayload = (event) => {
  const buildHookData = parseBuildHookData();
  const parsedEventBody = JSON.parse(event.body);
  // This is Netlify's default post-deployment payload. We include it with our custom data in case
  // we want to process any information
  const netlifyPayload = parsedEventBody.payload;
  return {
    netlifyPayload,
    ...buildHookData,
  };
};
