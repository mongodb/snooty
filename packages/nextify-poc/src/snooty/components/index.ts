import Paragraph from "./Paragraph";
import ComponentFactory from "./ComponentFactory";
export { Paragraph, ComponentFactory };
