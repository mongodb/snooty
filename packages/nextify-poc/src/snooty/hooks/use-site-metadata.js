import { useStaticQuery, graphql } from "../../gatsby-shim";

export const useSiteMetadata = () => {
  const { site } = useStaticQuery(
    graphql`
      query SiteMetaData {
        site {
          siteMetadata {
            commitHash
            database
            parserBranch
            parserUser
            patchId
            pathPrefix
            project
            reposDatabase
            siteUrl
            snootyBranch
            snootyEnv
            user
          }
        }
      }
    `
  );
  return site.siteMetadata;
};
