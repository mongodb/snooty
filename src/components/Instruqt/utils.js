export const getLabURL = (embedValue) => {
  return `https://play.instruqt.com/embed${embedValue}`;
};
