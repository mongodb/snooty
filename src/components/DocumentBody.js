import React from 'react';
import PropTypes from 'prop-types';
import ComponentFactory from './ComponentFactory';
import { getNestedValue } from '../utils/get-nested-value';

const DocumentBody = ({ addPillstrip, footnotes, metadata, pillstrips, refDocMapping, slug }) => {
  const pageNodes = getNestedValue(['ast', 'children'], refDocMapping) || [];
  return (
    <React.Fragment>
      {pageNodes.map((child, index) => (
        <ComponentFactory
          addPillstrip={addPillstrip}
          footnotes={footnotes}
          key={index}
          metadata={metadata}
          nodeData={child}
          refDocMapping={refDocMapping}
          pillstrips={pillstrips}
          slug={slug}
        />
      ))}
    </React.Fragment>
  );
};

DocumentBody.propTypes = {
  addPillstrip: PropTypes.func,
  footnotes: PropTypes.objectOf(PropTypes.object),
  pillstrips: PropTypes.objectOf(PropTypes.object),
  refDocMapping: PropTypes.shape({
    ast: PropTypes.shape({
      children: PropTypes.array,
    }).isRequired,
  }).isRequired,
  slug: PropTypes.string.isRequired,
};

DocumentBody.defaultProps = {
  addPillstrip: () => {},
  footnotes: {},
  pillstrips: {},
};

export default DocumentBody;
