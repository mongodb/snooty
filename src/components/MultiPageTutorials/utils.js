import { reportAnalytics } from '../../utils/report-analytics';

export const reportMPTAnalytics = (targetSlug, variant) => {
  reportAnalytics('MultiPageTutorialNextClicked', {
    targetSlug,
    variant,
  });
};
