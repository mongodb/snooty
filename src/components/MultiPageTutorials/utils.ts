import { reportAnalytics } from '../../utils/report-analytics';

export const reportMPTAnalytics = (targetSlug: string, variant: string) => {
  reportAnalytics('MultiPageTutorialNextClicked', {
    targetSlug,
    variant,
  });
};
