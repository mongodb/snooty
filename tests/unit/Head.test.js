import React from 'react';
import { render, screen } from '@testing-library/react';
import { Head } from '../../src/components/DocumentBody';
import mockStaticQuery from '../utils/mockStaticQuery';
import { useSiteMetadata } from '../../src/hooks/use-site-metadata';
import { generatePrefix } from '../../src/components/VersionDropdown/utils';
import useSnootyMetadata from '../../src/utils/use-snooty-metadata';
import { AVAILABLE_LANGUAGES } from '../../src/utils/locale';
import { DOTCOM_BASE_URL } from '../../src/utils/base-url';
import mockCompleteEOLPageContext from './data/CompleteEOLPageContext.json';
import mockEOLSnootyMetadata from './data/EOLSnootyMetadata.json';
import mockHeadPageContext from './data/HeadPageContext.test.json';

jest.mock(`../../src/utils/use-snooty-metadata`, () => jest.fn());

describe('Head', () => {
  beforeEach(() => {
    mockStaticQuery({
      siteUrl: DOTCOM_BASE_URL,
    });
  });

  describe("Canonical for completely EOL'd", () => {
    beforeEach(() => {
      useSnootyMetadata.mockImplementation(() => mockEOLSnootyMetadata);
    });
    it('renders the canonical tag from the snooty.toml', () => {
      render(<Head pageContext={mockCompleteEOLPageContext} />);
      const _canonical = mockEOLSnootyMetadata.canonical;
      const canonicalTag = screen.getByTestId('canonical');
      expect(canonicalTag).toBeInTheDocument();
      expect(canonicalTag).toHaveAttribute('id', 'canonical');
      expect(canonicalTag).toHaveAttribute('rel', 'canonical');
      expect(canonicalTag).toHaveAttribute('href', _canonical);
    });
  });

  describe("Canonical for EOL'd version", () => {
    beforeEach(() => {
      useSnootyMetadata.mockImplementation(() => mockEOLSnootyMetadata);
    });
    it('renders the canonical tag structured from the Head component with trailing slash', () => {
      render(<Head pageContext={mockHeadPageContext} />);
      const { siteUrl } = useSiteMetadata();
      const urlSlug = 'stable';
      const siteBasePrefix = mockHeadPageContext.repoBranches.siteBasePrefix;

      const currentVersion = `${siteUrl}/${siteBasePrefix}/${urlSlug}/`;

      const canonicalTag = screen.getByTestId('canonical');
      expect(canonicalTag).toBeInTheDocument();
      expect(canonicalTag).toHaveAttribute('id', 'canonical');
      expect(canonicalTag).toHaveAttribute('rel', 'canonical');
      expect(canonicalTag).toHaveAttribute('href', currentVersion);
    });
  });

  describe("Canonical for non-EoL'd", () => {
    beforeEach(() => {
      const modMockEOLSnootyMetadataToBeNotEOL = { ...mockEOLSnootyMetadata, eol: false };
      useSnootyMetadata.mockImplementation(() => modMockEOLSnootyMetadataToBeNotEOL);
    });

    it('renders the canonical tag that points to itself', () => {
      render(<Head pageContext={mockHeadPageContext} />);
      const siteMetadata = useSiteMetadata();
      const { siteUrl, parserBranch } = siteMetadata;
      const urlSlug = mockHeadPageContext.repoBranches.branches.find(
        (branch) => branch.gitBranchName === parserBranch
      )?.urlSlug;
      const pathPrefix = generatePrefix(urlSlug, siteMetadata);
      const slug = mockHeadPageContext.slug;

      const canonical = `${siteUrl}${pathPrefix}/${slug === '/' ? '' : slug}`;

      const canonicalTag = screen.getByTestId('canonical');
      expect(canonicalTag).toBeInTheDocument();
      expect(canonicalTag).toHaveAttribute('id', 'canonical');
      expect(canonicalTag).toHaveAttribute('rel', 'canonical');
      expect(canonicalTag).toHaveAttribute('href', canonical);
    });
  });

  describe('Canonical when pulled from directive', () => {
    beforeEach(() => {
      useSnootyMetadata.mockImplementation(() => mockEOLSnootyMetadata);
    });

    const metaCanonical = {
      type: 'directive',
      children: [],
      domain: '',
      name: 'meta',
      argument: [],
      options: {
        canonical: 'http://we-the-best.com',
      },
    };

    const expectCanonicalTagToBeCorrect = (canonicalTag) => {
      expect(canonicalTag).toBeInTheDocument();
      expect(canonicalTag).toHaveAttribute('id', 'canonical');
      expect(canonicalTag).toHaveAttribute('rel', 'canonical');
      expect(canonicalTag).toHaveAttribute('href', metaCanonical.options.canonical);
    };

    let mockPageContext = mockHeadPageContext;
    it('renders the canonical tag from directive rather than pulling from snooty.toml', () => {
      mockPageContext = mockCompleteEOLPageContext;
      mockPageContext.page.children.push(metaCanonical);
      render(<Head pageContext={mockPageContext} />);

      const canonicalTag = screen.getByTestId('canonical');
      expectCanonicalTagToBeCorrect(canonicalTag);
    });

    it("renders the canonical tag from directive rather than pulling from stable branch (version eol'd)", () => {
      mockPageContext.page.children.push(metaCanonical);
      render(<Head pageContext={mockPageContext} />);

      const canonicalTag = screen.getByTestId('canonical');
      expectCanonicalTagToBeCorrect(canonicalTag);
    });

    it('renders the canonical tag from directive rather than pulling from default logic', () => {
      //need to override what happens in the beforeEach of this describe
      const modMockEOLSnootyMetadataToBeNotEOL = { ...mockEOLSnootyMetadata, eol: false };
      useSnootyMetadata.mockImplementation(() => modMockEOLSnootyMetadataToBeNotEOL);

      mockPageContext.page.children.push(metaCanonical);
      render(<Head pageContext={mockPageContext} />);

      const canonicalTag = screen.getByTestId('canonical');
      expectCanonicalTagToBeCorrect(canonicalTag);
    });
  });

  describe('hreflang links', () => {
    beforeEach(() => {
      useSnootyMetadata.mockImplementation(() => ({ ...mockEOLSnootyMetadata, eol: false }));
    });

    it.each([['/'], ['foo']])('renders based on slug', (slug) => {
      const mockPageContext = { ...mockHeadPageContext, slug };
      const { container } = render(<Head pageContext={mockPageContext} />);
      const hrefLangLinks = container.querySelectorAll('link.sl_opaque');
      expect(hrefLangLinks).toHaveLength(AVAILABLE_LANGUAGES.length);
      expect(hrefLangLinks).toMatchSnapshot();
    });
  });
});
