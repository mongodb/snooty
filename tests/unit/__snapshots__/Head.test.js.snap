// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Head hreflang links renders based on slug 1`] = `
NodeList [
  <link
    class="sl_opaque"
    href="https://www.mongodb.com/"
    hreflang="x-default"
    rel="alternate"
  />,
  <link
    class="sl_opaque"
    href="https://www.mongodb.com/zh-cn/"
    hreflang="zh-cn"
    rel="alternate"
  />,
  <link
    class="sl_opaque"
    href="https://www.mongodb.com/ko-kr/"
    hreflang="ko-kr"
    rel="alternate"
  />,
  <link
    class="sl_opaque"
    href="https://www.mongodb.com/pt-br/"
    hreflang="pt-br"
    rel="alternate"
  />,
]
`;

exports[`Head hreflang links renders based on slug 2`] = `
NodeList [
  <link
    class="sl_opaque"
    href="https://www.mongodb.com/foo/"
    hreflang="x-default"
    rel="alternate"
  />,
  <link
    class="sl_opaque"
    href="https://www.mongodb.com/zh-cn/foo/"
    hreflang="zh-cn"
    rel="alternate"
  />,
  <link
    class="sl_opaque"
    href="https://www.mongodb.com/ko-kr/foo/"
    hreflang="ko-kr"
    rel="alternate"
  />,
  <link
    class="sl_opaque"
    href="https://www.mongodb.com/pt-br/foo/"
    hreflang="pt-br"
    rel="alternate"
  />,
]
`;
